File: server/src/main/kotlin/com/blueoauld/server/member/application/MemberImageService.kt
package com.blueoauld.server.member.application

import com.blueoauld.server.r2.application.R2Provider
import com.blueoauld.server.r2.application.request.UploadUrlRequests
import com.blueoauld.server.r2.application.response.UploadUrlResponses
import com.blueoauld.server.r2.type.FileContentType
import org.springframework.stereotype.Service
import java.time.Duration
import java.util.*

@Service
class MemberImageService(

    private val r2Provider: R2Provider,
) {

    fun createPublicUploadUrls(requests: UploadUrlRequests): UploadUrlResponses {
        val urls = requests.urls.map {
            val contentType = FileContentType.from(it.contentType)
            val fileName = "${UUID.randomUUID()}.${contentType.extension}"
            val key = "member/public/temporary/$fileName"

            r2Provider.createUploadUrl(key, it.contentType, Duration.ofMinutes(5))
        }
        return UploadUrlResponses(urls)
    }

    fun createPrivateUploadUrls(requests: UploadUrlRequests): UploadUrlResponses {
        val urls = requests.urls.map {
            val contentType = FileContentType.from(it.contentType)
            val fileName = "${UUID.randomUUID()}.${contentType.extension}"
            val key = "member/private/temporary/$fileName"

            r2Provider.createUploadUrl(key, it.contentType, Duration.ofMinutes(5))
        }
        return UploadUrlResponses(urls)
    }
}

File: app/app/Service/Member/MemberImageService.swift
import Alamofire

final class MemberImageService {

    static let shared = MemberImageService()

    func createPublicUploadUrls(
        urls: UploadUrlRequests
    ) async throws -> UploadUrlResponses {
        try await PrivateNetworkManager.shared.request(
            "/members/images/public/urls",
            method: .post,
            parameters: urls,
            encoder: JSONParameterEncoder.default,
            as: UploadUrlResponses.self
        )
    }

    func createPrivateUploadUrls(
        urls: UploadUrlRequests
    ) async throws -> UploadUrlResponses {
        try await PrivateNetworkManager.shared.request(
            "/members/images/private/urls",
            method: .post,
            parameters: urls,
            encoder: JSONParameterEncoder.default,
            as: UploadUrlResponses.self
        )
    }
}
